{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "0a32996d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Imports\n",
    "import random"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "93af1ca8",
   "metadata": {},
   "source": [
    "## Helper functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "da680914",
   "metadata": {},
   "outputs": [],
   "source": [
    "def ReLU(x):\n",
    "    if x < 0:\n",
    "        return 0\n",
    "    return x\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "35389618",
   "metadata": {},
   "source": [
    "## NEAT Algorithm\n",
    "\n",
    "Original paper: https://nn.cs.utexas.edu/downloads/papers/stanley.ec02.pdf\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e86b92b7",
   "metadata": {},
   "source": [
    "#### Network Encoding\n",
    "\n",
    "In the NEAT algorithm each neuron in the neural network is represented as:\n",
    "\n",
    "![image](assets/genotype.png)\n",
    "\n",
    "*Source: [Evolving Neural Networks through Augmenting Topologies](https://nn.cs.utexas.edu/downloads/papers/stanley.ec02.pdf)*\n",
    "\n",
    "\n",
    "#### Mutation\n",
    "\n",
    "\n",
    "There are 2 types of mutation:\n",
    "- Add connection\n",
    "- Add node\n",
    "\n",
    "![image](assets/mutation.png)\n",
    "\n",
    "*Source: [Evolving Neural Networks through Augmenting Topologies](https://nn.cs.utexas.edu/downloads/papers/stanley.ec02.pdf)*\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb0bea0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "class InnovationNumber:\n",
    "    def __init__(self):\n",
    "        self.value = 0\n",
    "    \n",
    "    def pop(self):\n",
    "        \"\"\" Returns the innovation number and increments it automatically \"\"\"\n",
    "        tmp = self.value\n",
    "        self.value += 1\n",
    "        return tmp\n",
    "\n",
    "class NodeGene:\n",
    "    def __init__(self, layer, activation, bias):\n",
    "        self.layer = layer  # The layer to which the node belongs\n",
    "        self.activation = activation    # Activation function\n",
    "        self.bias = bias\n",
    "\n",
    "class ConncectionGene:\n",
    "    def __init__(self, in_node: NodeGene, out_node: NodeGene, weight: float,  innov: int, enabled: bool = True):\n",
    "        self.in_node = in_node\n",
    "        self.out_node = out_node\n",
    "        self.weight = weight\n",
    "        self.enabled = enabled  # Whether the connection is enabled or not\n",
    "        self.innov = innov  # Innovation number described in the paper\n",
    "    \n",
    "    def copy(self):\n",
    "        return ConncectionGene(self.in_node, self.out_node, self.weight, self.enabled, self.innov)\n",
    "\n",
    "class Genome:\n",
    "    def __init__(self, nodes, connections):\n",
    "        self.nodes = nodes\n",
    "        self.connections = connections\n",
    "    \n",
    "    def mutate_add_connection(self, innov: InnovationNumber):\n",
    "        node1, node2 = random.sample(self.nodes, 2)\n",
    "        if node1 == node2:\n",
    "            return\n",
    "        for c in self.connections:\n",
    "            if c.in_node == node1 and c.out_node == node2:\n",
    "                return\n",
    "        \n",
    "        new_conn = ConncectionGene(node1, node2, random.uniform(-1, 1), innov.pop(), True)\n",
    "        self.connections.append(new_conn)\n",
    "    \n",
    "    def mutate_add_node(self, innov: InnovationNumber):\n",
    "        connection = random.sample(self.connections)\n",
    "        connection.enabled = False  # Disable the connection\n",
    "\n",
    "        new_node = NodeGene(\"hid\", ReLU, random.uniform(-1,1))\n",
    "\n",
    "        conn1 = ConncectionGene(connection.in_node, new_node, 1, innov.pop(), True)\n",
    "        conn2 = ConncectionGene(new_node, connection.out_node, connection.weight, innov.pop(), True)\n",
    "\n",
    "        self.nodes.append(new_node)\n",
    "        self.connections.extend([conn1, conn2])\n",
    "\n",
    "\n",
    "def find_disjoint(set1, set2):\n",
    "    set1_dict = {item.innov: item for item in set1}\n",
    "    set2_dict = {item.innov: item for item in set2}\n",
    "\n",
    "    disjoint = []\n",
    "\n",
    "    # Innovations in set1 but not in set2\n",
    "    for innov, item in set1_dict.items():\n",
    "        if innov not in set2_dict:\n",
    "            disjoint.append(item)\n",
    "\n",
    "    # Innovations in set2 but not in set1\n",
    "    for innov, item in set2_dict.items():\n",
    "        if innov not in set1_dict:\n",
    "            disjoint.append(item)\n",
    "\n",
    "    return disjoint\n",
    "\n",
    "\n",
    "def crossover(parent1: Genome, parent2: Genome) -> Genome:\n",
    "    \"\"\" Crossover assuming parent1 is the fittest parent \"\"\"\n",
    "    # Build maps of genes keyed by innovation number\n",
    "    genes1 = {g.innov: g for g in parent1.connections}\n",
    "    genes2 = {g.innov: g for g in parent2.connections}\n",
    "\n",
    "    offspring_connections = []\n",
    "    offspring_nodes = set()\n",
    "\n",
    "    # Combine all innovation numbers\n",
    "    all_innovs = set(genes1.keys()) | set(genes2.keys())\n",
    "\n",
    "    for innov in sorted(all_innovs):\n",
    "        gene1 = genes1.get(innov)\n",
    "        gene2 = genes2.get(innov)\n",
    "        \n",
    "        if gene1 and gene2:  # Matching genes\n",
    "            selected = random.choice([gene1, gene2])\n",
    "            gene_copy = selected.copy()\n",
    "\n",
    "        elif gene1 and not gene2:   # Disjoint gene (from the fittest parent)\n",
    "            gene_copy = gene1.copy()\n",
    "        \n",
    "        else:   # Not taking disjoint genes from less fit parent\n",
    "            continue\n",
    "\n",
    "        offspring_connections.append(gene_copy)\n",
    "        # offspring_nodes.extend([gene_copy.in_node, gene_copy.out_node])\n",
    "        offspring_nodes.add(gene_copy.in_node)\n",
    "        offspring_nodes.add(gene_copy.out_node)\n",
    "\n",
    "    offspring_nodes = list(offspring_nodes) # Remove the duplicates\n",
    "    return Genome(offspring_nodes, offspring_connections)\n",
    "    \n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
