{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0a32996d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Imports\n",
    "import random\n",
    "import math"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "93af1ca8",
   "metadata": {},
   "source": [
    "## Helper functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "da680914",
   "metadata": {},
   "outputs": [],
   "source": [
    "def ReLU(x):\n",
    "    if x < 0:\n",
    "        return 0\n",
    "    return x\n",
    "\n",
    "def sigmoid(x):\n",
    "    return 1/(1 + math.exp(-x))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "35389618",
   "metadata": {},
   "source": [
    "## NEAT Algorithm\n",
    "\n",
    "Original paper: https://nn.cs.utexas.edu/downloads/papers/stanley.ec02.pdf\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e86b92b7",
   "metadata": {},
   "source": [
    "#### Network Encoding\n",
    "\n",
    "In the NEAT algorithm each neuron in the neural network is represented as:\n",
    "\n",
    "![image](assets/genotype.png)\n",
    "\n",
    "*Source: [Evolving Neural Networks through Augmenting Topologies](https://nn.cs.utexas.edu/downloads/papers/stanley.ec02.pdf)*\n",
    "\n",
    "\n",
    "#### Mutation\n",
    "\n",
    "There are 2 types of mutation:\n",
    "- Add connection\n",
    "- Add node\n",
    "\n",
    "![image](assets/mutation.png)\n",
    "\n",
    "*Source: [Evolving Neural Networks through Augmenting Topologies](https://nn.cs.utexas.edu/downloads/papers/stanley.ec02.pdf)*\n",
    "\n",
    "\n",
    "#### Crossover\n",
    "\n",
    "![image](assets/mutation.png)\n",
    "\n",
    "*Source: [Evolving Neural Networks through Augmenting Topologies](https://nn.cs.utexas.edu/downloads/papers/stanley.ec02.pdf)*\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "eb0bea0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "class InnovationNumber:\n",
    "    def __init__(self):\n",
    "        self.value = 0\n",
    "    \n",
    "    def pop(self):\n",
    "        \"\"\" Returns the innovation number and increments it automatically \"\"\"\n",
    "        tmp = self.value\n",
    "        self.value += 1\n",
    "        return tmp\n",
    "\n",
    "class NodeGene:\n",
    "    def __init__(self, layer, activation, bias):\n",
    "        self.layer = layer  # The layer to which the node belongs\n",
    "        self.activation = activation    # Activation function\n",
    "        self.bias = bias\n",
    "\n",
    "class ConnectionGene:\n",
    "    def __init__(self, in_node: NodeGene, out_node: NodeGene, weight: float,  innov: int, enabled: bool = True):\n",
    "        self.in_node = in_node\n",
    "        self.out_node = out_node\n",
    "        self.weight = weight\n",
    "        self.enabled = enabled  # Whether the connection is enabled or not\n",
    "        self.innov = innov  # Innovation number described in the paper\n",
    "    \n",
    "    def copy(self):\n",
    "        return ConnectionGene(self.in_node, self.out_node, self.weight, self.enabled, self.innov)\n",
    "\n",
    "class Genome:\n",
    "    def __init__(self, nodes, connections):\n",
    "        self.nodes = nodes\n",
    "        self.connections = connections\n",
    "    \n",
    "    def mutate_add_connection(self, innov: InnovationNumber):\n",
    "        node1, node2 = random.sample(self.nodes, 2)\n",
    "        if node1 == node2:\n",
    "            return\n",
    "        for c in self.connections:\n",
    "            if c.in_node == node1 and c.out_node == node2:\n",
    "                return\n",
    "        \n",
    "        new_conn = ConnectionGene(node1, node2, random.uniform(-1, 1), innov.pop(), True)\n",
    "        self.connections.append(new_conn)\n",
    "    \n",
    "    def mutate_add_node(self, innov: InnovationNumber):\n",
    "        connection = random.sample(self.connections)\n",
    "        connection.enabled = False  # Disable the connection\n",
    "\n",
    "        new_node = NodeGene(\"hid\", ReLU, random.uniform(-1,1))\n",
    "\n",
    "        conn1 = ConnectionGene(connection.in_node, new_node, 1, innov.pop(), True)\n",
    "        conn2 = ConnectionGene(new_node, connection.out_node, connection.weight, innov.pop(), True)\n",
    "\n",
    "        self.nodes.append(new_node)\n",
    "        self.connections.extend([conn1, conn2])\n",
    "\n",
    "    def __topological_sort(self, edges):\n",
    "        visited = set()\n",
    "        order = []\n",
    "\n",
    "        def visit(n):\n",
    "            if n in visited:\n",
    "                return\n",
    "            visited.add(n)\n",
    "            for m in edges[n]:\n",
    "                visit(m)\n",
    "            order.append(n)\n",
    "\n",
    "        for node in edges:\n",
    "            visit(node)\n",
    "\n",
    "        return order[::-1]\n",
    "\n",
    "    def evaluate(self, input_values: list[float]):\n",
    "        node_values = {}\n",
    "        node_inputs = {n: [] for n in self.nodes}\n",
    "\n",
    "        input_nodes = [n for n in self.nodes if n.layer == \"input\"]\n",
    "        output_nodes = [n for n in self.nodes if n.layer == \"output\"]\n",
    "\n",
    "        # Verify that the number of input values matches the number of input nodes\n",
    "        if len(input_nodes) != len(input_values):\n",
    "            raise ValueError(\"Number of inputs doesn't match number of input nodes\")\n",
    "        \n",
    "        # Assign input values\n",
    "        for node, val in zip(input_nodes, input_values):\n",
    "            node_values[node] = val\n",
    "\n",
    "        edges = {}\n",
    "        for n in self.nodes:\n",
    "            edges[n] = []\n",
    "        \n",
    "        for conn in self.connections:\n",
    "            if conn.enabled:\n",
    "                edges[conn.in_node].append(conn.out_node)\n",
    "                node_inputs[conn.out_node].append(conn)\n",
    "        \n",
    "        sorted_nodes = self.__topological_sort(edges)\n",
    "\n",
    "        for node in sorted_nodes:\n",
    "            if node in node_values:\n",
    "                continue\n",
    "            \n",
    "            incoming = node_inputs[node]\n",
    "            total_input = sum(\n",
    "                node_values[c.in_node] * c.weight for c in incoming\n",
    "            ) + node.bias\n",
    "\n",
    "            node_values[node] = node.activation(total_input)\n",
    "\n",
    "        return [node_values[n] for n in output_nodes]\n",
    "\n",
    "\n",
    "def crossover(parent1: Genome, parent2: Genome) -> Genome:\n",
    "    \"\"\" Crossover assuming parent1 is the fittest parent \"\"\"\n",
    "    # Build maps of genes keyed by innovation number\n",
    "    genes1 = {g.innov: g for g in parent1.connections}\n",
    "    genes2 = {g.innov: g for g in parent2.connections}\n",
    "\n",
    "    offspring_connections = []\n",
    "    offspring_nodes = set()\n",
    "\n",
    "    # Combine all innovation numbers\n",
    "    all_innovs = set(genes1.keys()) | set(genes2.keys())\n",
    "\n",
    "    for innov in sorted(all_innovs):\n",
    "        gene1 = genes1.get(innov)\n",
    "        gene2 = genes2.get(innov)\n",
    "        \n",
    "        if gene1 and gene2:  # Matching genes\n",
    "            selected = random.choice([gene1, gene2])\n",
    "            gene_copy = selected.copy()\n",
    "\n",
    "        elif gene1 and not gene2:   # Disjoint gene (from the fittest parent)\n",
    "            gene_copy = gene1.copy()\n",
    "        \n",
    "        else:   # Not taking disjoint genes from less fit parent\n",
    "            continue\n",
    "\n",
    "        offspring_connections.append(gene_copy)\n",
    "        offspring_nodes.add(gene_copy.in_node)\n",
    "        offspring_nodes.add(gene_copy.out_node)\n",
    "\n",
    "    offspring_nodes = list(offspring_nodes) # Remove the duplicates\n",
    "    return Genome(offspring_nodes, offspring_connections)\n",
    "\n",
    "\n",
    "def distance(genome1: Genome, genome2: Genome, c1=1.0, c2=1.0, c3=0.4):\n",
    "    genes1 = {g.innov: g for g in genome1.connections}\n",
    "    genes2 = {g.innov: g for g in genome2.connections}\n",
    "\n",
    "    innovations1 = set(genes1.keys())\n",
    "    innovations2 = set(genes2.keys())\n",
    "\n",
    "    matching = innovations1 & innovations2\n",
    "    disjoint = (innovations1 ^ innovations2)\n",
    "    excess = set()\n",
    "\n",
    "    max_innov1 = max(innovations1) if innovations1 else 0\n",
    "    max_innov2 = max(innovations2) if innovations2 else 0\n",
    "    max_innov = min(max_innov1, max_innov2)\n",
    "\n",
    "    # Separate excess from disjoint\n",
    "    for innov in disjoint.copy():\n",
    "        if innov > max_innov:\n",
    "            excess.add(innov)\n",
    "            disjoint.remove(innov)\n",
    "\n",
    "    # Weight difference of matching genes\n",
    "    if matching:\n",
    "        weight_diff = sum(\n",
    "            abs(genes1[i].weight - genes2[i].weight) for i in matching\n",
    "        )\n",
    "        avg_weight_diff = weight_diff / len(matching)\n",
    "    else:\n",
    "        avg_weight_diff = 0\n",
    "\n",
    "    # Normalize by N\n",
    "    N = max(len(genome1.connections), len(genome2.connections))\n",
    "    if N < 20:  # NEAT typically uses 1 if N < 20\n",
    "        N = 1\n",
    "\n",
    "    delta = (c1 * len(excess)) / N + (c2 * len(disjoint)) / N + c3 * avg_weight_diff\n",
    "    return delta\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "a3f523f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Species:\n",
    "    def __init__(self, representative: Genome):\n",
    "        self.representative = representative\n",
    "        self.members = [representative]\n",
    "        self.adjusted_fitness = 0\n",
    "        self.best_fitness = -float('inf')\n",
    "        self.stagnant_generations = 0\n",
    "    \n",
    "    def add_member(self, genome: Genome):\n",
    "        self.members.append(genome)\n",
    "    \n",
    "    def clear_members(self):\n",
    "        self.members = []\n",
    "    \n",
    "    def compute_adjusted_fitness(self, fitnesses: dict):\n",
    "        self.adjusted_fitness = 0\n",
    "        for genome in self.members:\n",
    "            self.adjusted_fitness += fitnesses[genome] / len(self.members)\n",
    "\n",
    "\n",
    "class Speciator:\n",
    "    def __init__(self, compatibility_threshold=3.0):\n",
    "        self.species = []\n",
    "        self.compatibility_threshold = compatibility_threshold\n",
    "    \n",
    "    def speciate(self, population: list[Genome], fitnesses: dict):\n",
    "        \"\"\" Group genomes into species based on distance \"\"\"\n",
    "        # Clear all species for the new generation\n",
    "        for s in self.species:\n",
    "            s.clear_members()\n",
    "        \n",
    "        for genome in population:\n",
    "            found_species = False\n",
    "            for species in self.species:\n",
    "                if distance(genome, species.representative) < self.compatibility_threshold:\n",
    "                    species.add_member(genome)\n",
    "                    found_species = True\n",
    "                    break\n",
    "            \n",
    "            if not found_species:\n",
    "                new_species = Species(representative=genome)\n",
    "                self.species.append(new_species)\n",
    "\n",
    "        # Remove empty species\n",
    "        self.species = [s for s in self.species if s.members]\n",
    "\n",
    "        # Recompute adjusted fitness\n",
    "        for s in self.species:\n",
    "            s.compute_adjusted_fitness(fitnesses)\n",
    "\n",
    "    def get_species(self):\n",
    "        return self.species\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "df682d5c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Genome output: [0.6737025243079816]\n"
     ]
    }
   ],
   "source": [
    "input1 = NodeGene(\"input\", activation=lambda x: x, bias=0)\n",
    "input2 = NodeGene(\"input\", activation=lambda x: x, bias=0)\n",
    "hidden = NodeGene(\"hid\", activation=sigmoid, bias=0.1)\n",
    "output = NodeGene(\"output\", activation=sigmoid, bias=0.2)\n",
    "\n",
    "conn1 = ConnectionGene(input1, hidden, weight=0.5, innov=1)\n",
    "conn2 = ConnectionGene(input2, hidden, weight=0.5, innov=2)\n",
    "conn3 = ConnectionGene(hidden, output, weight=1.0, innov=3)\n",
    "\n",
    "nodes = [input1, input2, hidden, output]\n",
    "connections = [conn1, conn2, conn3]\n",
    "genome = Genome(nodes, connections)\n",
    "\n",
    "result = genome.evaluate([1.0, -1.0])\n",
    "print(\"Genome output:\", result)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
